pyuoi/linear_model/poisson.py

```diff
@@ -1,45 +1,471 @@
         self.max_iter = max_iter
         self.tol = tol
+        self.warm_start = warm_start
+
+    def fit(self, X, y, init=None):
+        """Fit the Poisson GLM with coordinate descent.
+
+        Parameters
+        ----------
+        X : nd-array, shape (n_samples, n_features)
+            The design matrix.
+
+        y : nd-array, shape (n_samples,)
+            Response vector. Will be cast to X's dtype if necessary.
+            Currently, this implementation does not handle multiple response
+            variables.
 
-    def fit(self, X, y):
+        init : nd-array, shape (n_features)
+            Initialization for parameters.
+        """
         self.n_samples, self.n_features = X.shape
 
         # initialization
-        beta0 = np.log(np.mean(y))
-        beta = np.zeros(n_features)
+        if self.warm_start and hasattr(self, 'coef_'):
+            coef = self.coef_
+        else:
+            coef = np.random.normal(size=(self.n_features))
+
+        if init is not None:
+            coef = init
+
+
+        if self.fit_intercept:
+            intercept = np.random.normal()
+        else:
+            intercept = 0
+
+        # we will handle the intercept by hand: only preprocess the design
+        # matrix
+        X, _, X_offset, _, X_scale = _preprocess_data(
+            X, y, fit_intercept=False, normalize=self.normalize)
+
+        # all features are initially active
+        active_idx = np.arange(self.n_features)
+
+        coef_update = np.zeros(coef.shape)
+        # perform coordinate descent updates
+        for iteration in range(self.max_iter):
+
+            # linearize the log-likelihood
+            w, z = self.adjusted_response(X, y, coef, intercept)
+
+            # perform an update of coordinate descent
+            coef_update, intercept = self.cd_sweep(
+                coef=coef, intercept=intercept, X=X, w=w, z=z,
+                active_idx=active_idx)
+
+            # check convergence
+            if np.max(np.abs(coef_update - coef)) < self.tol:
+                break
+
+            coef = coef_update
+
+            # update the active features
+            active_idx = np.argwhere(coef != 0).ravel()
+
+        self.intercept_ = intercept
+        self.coef_ = coef_update / X_scale
+
+    def predict(self, X):
+        """Predict using the poisson generalized linear model. In this case,
+        the conditional means are returned.
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+            Design matrix to predict on.
 
-        # initially set all parameters to be active
-        active_idx = np.ones(n_features)
+        Returns
+        -------
+        mu : array_like, shape (n_samples)
+            The predicted response values, i.e. the conditional means.
+        """
+        if hasattr(self, 'coef_') and hasattr(self, 'intercept_'):
+            mu = np.exp(self.intercept_ + np.dot(X, self.coef_))
+            return mu
```

Comment:
I don't think this is consistent with how `predict` is used in sklearn. I think `predict` typically outputs the mode of the distribution. I'm not sure how other packages handle the poisson case.

For [logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict) `predict` outputs the class value with the highest value, not a vector of probabilities for each class.

For the poisson case, the corresponding mode would be

```suggestion
            return np.floor(mu).astype(int)
```

Reply:
This could be `predict_mean`

Reply:
This makes sense, but something else will have to be worked out. I put this in so that no adjustments had to be made in the base fit method for UoI, which relies on the predict() method to obtain scores for comparing models. 

For Poisson, the likelihood is used, and only interacts with mu = log(X*beta), not the mode.